### ISSUE TITLE
Crash on backwards step when using `batch_first=True` for LSTMs on MPS (1.14 nightly build)

### ISSUE THREAD

**Issue by a2aaron** [pytorch/pytorch#87579]
### 🐛 Describe the bug

When using `batch_first=True` for LSTM layers, attempting to compute gradients causes a crash when running pytorch on MPS. This crash does not happen when running on CPU.

```python
import torch
import torch.nn as nn

device = "mps" if torch.backends.mps.is_available() else "cpu"
print("Running on", device)

data = torch.zeros((5, 3, 10), device=device)

lstm = nn.LSTM(10, 20, 1, batch_first=True).to(device)
h0 = torch.zeros((1, 5, 20), device=device)
c0 = torch.zeros((1, 5, 20), device=device)

expected = torch.randn((5, 3, 20), device=device)

output, _ = lstm(data, (h0, c0))
output = output.sum()

# crash occurs here
output.backward()
```

This produces the following crash, when running on MPS:

```
loc("total derivative last state"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/a0876c02-1788-11ed-b9c4-96898e02b808/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":219:0)): error: input types 'tensor<1x3x20xf32>' and 'tensor<1x5x20xf32>' are not broadcast compatible
LLVM ERROR: Failed to infer result type(s).
```

Additionally, I searched through the other issues, it seems there is a closed issue [here](https://github.com/pytorch/pytorch/issues/80306) which has a similar bug. However, despite being closed, it seems like it is still broken

### Versions

```
Collecting environment information...
PyTorch version: 1.14.0.dev20221023
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.102)
CMake version: version 3.24.1
Libc version: N/A

Python version: 3.10.7 (main, Sep 14 2022, 22:38:23) [Clang 14.0.0 (clang-1400.0.29.102)] (64-bit runtime)
Python platform: macOS-12.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.4
[pip3] pytorch-lightning==1.4.2
[pip3] torch==1.14.0.dev20221023
[pip3] torch-fidelity==0.3.0
[pip3] torchaudio==0.14.0.dev20221023
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.15.0.dev20221023
[conda] Could not collect
```

cc @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by hvaara** [comment]
I tried running the reproducer, and it doesn't crash anymore.

```python
import torch
import torch.nn as nn

device = "mps" if torch.backends.mps.is_available() else "cpu"
print("Running on", device)

data = torch.zeros((5, 3, 10), device=device)

lstm = nn.LSTM(10, 20, 1, batch_first=True).to(device)
h0 = torch.zeros((1, 5, 20), device=device)
c0 = torch.zeros((1, 5, 20), device=device)

expected = torch.randn((5, 3, 20), device=device)

output, _ = lstm(data, (h0, c0))
output = output.sum()

# crash occurs here
output.backward()

# Output:
# Running on mps
```

I believe this issue has been fixed. Please reopen if there are still items to be addressed.

@pytorchbot close
